### PR TITLE
Zonestar ZM3E2, ZM3E4 V1, ZM3E4 V2

### PR DESCRIPTION
**Pull request by ellensp**
### Description

As requested in https://github.com/MarlinFirmware/Marlin/issues/22492
Add ZONESTAR boards
BOARD_ZONESTAR_ZM3E2
BOARD_ZONESTAR_ZM3E4
BOARD_ZONESTAR_ZM3E4V2

Information extracted from provided zip in above FR

### Requirements

One of these controllers

### Benefits

These controllers build 

### Related Issues
https://github.com/MarlinFirmware/Marlin/issues/22492

### Important
This is just a straight extraction from the zip with very minor modifications and some formatting.
They build. But they use maple.
I have no hardware to test these on

### CODE REVIEW

**Comment by ellensp** [comment]
The simplification (which makes it much harder to read)  broke things 
BOARD_ZONESTAR_ZM3E2 ->  invalid length for memory region ram, during link
BOARD_ZONESTAR_ZM3E4 ->  invalid length for memory region ram, during link
BOARD_ZONESTAR_ZM3E4V2 -> invalid length for memory region ram, during link

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+#define BOARD_INFO_NAME "ZONESTAR ZM3E4 V2.0"
+
+//#define DISABLE_DEBUG
+#define DISABLE_JTAG
+
+#define FLASH_EEPROM_EMULATION
+#define EEPROM_PAGE_SIZE     (0x800) // 2KB
+#define EEPROM_START_ADDRESS uint32(0x8000000 + (STM32_FLASH_SIZE) * 1024 - 2 * EEPROM_PAGE_SIZE)
+#define E2END                (EEPROM_PAGE_SIZE - 1)
```
Comment (thisiskeithb) [inline Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: This (and other pins) should use the new `MARLIN_EEPROM_SIZE` convention for EEPROM.

Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: Done

**Comment by thinkyhead** [comment]
> The simplification (which makes it much harder to read) broke things

Good to know. I'm experimenting to see what works and what doesn't. If the `custom.ld` can be made to work that would allow dropping several other `.ld` files. But, "can it be made to work?" is the question.

I worked on this PR for a couple of hours earlier, but I still have to fully scan, optimize, and update the pins files, which are based on older pins file standards (from over a year ago).

**Comment by thinkyhead** [comment]
I worked on this PR for a couple of hours earlier, but I still have to fully scan, optimize, and update the pins files, which are based on older pins file standards (from over a year ago).

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+  #define EEPROM_START_ADDRESS  (0x08000000 + (STM32_FLASH_SIZE) * 1024 - 2 * EEPROM_PAGE_SIZE)
+  #define MARLIN_EEPROM_SIZE    EEPROM_PAGE_SIZE  // 2KB
+#endif
+
+//#define OPTION_DUALZ_DRIVE
+//#define OPTION_Z2_ENDSTOP
+//#define SWITCH_EXTRUDER_SEQUENCE
+
+//=============================================================================
+// ZONESTAR ZM3E4 V1.0 (STM32F130RCT6) board pin assignments
```
Comment (tpruvot) [inline Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h]: small typo here, STM32F130VCT6 i guess

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  #define MARLIN_EEPROM_SIZE    EEPROM_PAGE_SIZE  // 2KB
+#endif
+
+//#define OPTION_DUALZ_DRIVE
+//#define OPTION_Z2_ENDSTOP
+//#define OPTION_REPEAT_PRINTING
+//#define SWITCH_EXTRUDER_SEQUENCE
+
+//=============================================================================
+// ZONESTAR ZM3E4 V1.0 (STM32F130RCT6) board pin assignments
```
Comment (tpruvot) [inline Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: and here, STM32F103VET6 ?

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+#define BOARD_INFO_NAME "ZONESTAR ZM3E4 V2.0"
+
+// =====================NOTE=====================
+// #define TONE_TIMER     1
+// #define TONE_CHANNEL   2
+// Please redefine the TONE_TIMER and TONE_CHANNEL in
+// .platformio\packages\framework-arduinoststm32-maple\STM32F1\cores\maple\tone.cpp
+// line 31 and line      34
+// =====================NOTE=====================
```
Comment (tpruvot) [inline Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: you can define those in env build_flags -DTONE_TIMER=1 -DTONE_CHANNEL=2

https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/cores/maple/tone.cpp#L30-L46

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h (hunk 2 above)
Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h]: they are all 103's  a dyslexic copy and paste in supplied source

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h (hunk 3 above)
Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: Im guessing that too, but link files is confused
MEMORY
{
  ram (rwx) : ORIGIN = 0x20000000, LENGTH = 48K - 40
  rom (rx)  : ORIGIN = 0x08005000, LENGTH = 512K - 20K - 4K
}

512 flash/48k ram?  no such beast  I really hope they arent over flashing a VC

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h (hunk 2 above)
Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h]: changed

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h (hunk 3 above)
Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: changed (wont be able to resolve if its a bad lk or a over flashed board till get to see one)

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h (hunk 2 above)
Comment (tpruvot) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V1_0.h]: lol yea, and i didnt even notice, i copy pasted :p

File: Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h (hunk 4 above)
Comment (ellensp) [reply Marlin/src/pins/stm32f1/pins_ZM3E4_V2_0.h]: Updated

File: ini/stm32f1.ini [hunk 5]
```diff
@@ -374,3 +374,43 @@ extra_scripts        = ${stm32_variant.extra_scripts}
...
+extends              = ZONESTAR_ZM3E
+board                = genericSTM32F103RC
+board_build.variant  = MARLIN_F103Rx
+
+[env:STM32F103VC_ZM3E4_USB]
+platform             = ${ZONESTAR_ZM3E.platform}
+extends              = ZONESTAR_ZM3E
+board                = genericSTM32F103VC
+board_build.variant  = MARLIN_F103Vx
+build_flags          = ${ZONESTAR_ZM3E.build_flags} -DTONE_TIMER=1 -DTONE_CHANNEL=2
```
Comment (tpruvot) [inline ini/stm32f1.ini]: these defines are for maple only... STM32 use -DTIMER_TONE (not TONE_TIMER) ... no channel, then just a pin
https://github.com/stm32duino/Arduino_Core_STM32/blob/master/cores/arduino/Tone.cpp

Comment (thinkyhead) [reply ini/stm32f1.ini:408]: I see. Just being contrarian, are they? What's the other one… "`-DTHE_CHANNEL_FOR_THE_TONE_TIMER`" ?

Comment (thinkyhead) [reply ini/stm32f1.ini]: I see that `pins_BTT_SKR_MINI_E3_V2_0.h` also contains …
```cpp
#define BOARD_CUSTOM_BUILD_FLAGS -DTONE_CHANNEL=4 -DTONE_TIMER=4
```
I assume that this also only compatible with Maple, and is ignored on the non-Maple platforms.

**Pull request merged**